### ISSUE TITLE
Stop button not triggering TearDown

### ISSUE THREAD

**Issue by MartinZec** [TestCentric/testcentric-gui#374]
Hello,

I saw that the Stop button is working since Beta4.
In NUnit2, after using the Stop button, the test was stopped and the Cleanup and TearDown was called.
In Beta4 the test is stopped but no TearDown or OneTimeTearDown is called.

Since the TearDown is not called, the test context is not clean and I cannot start another test after pressing the Stop button.

Is there a different/new way to clean up a test in NUnit3?

Best regards,
Martin

**Comment by CharliePoole** [comment]
This was changed in issue #337.

What's interesting is that @JeffKryzer who wrote the issue commented that V2 worked differently than what you are saying. Thinking about it, I guess you are both right...

In V2, the test method was stopped forcibly __but__ the teardown was then run. In NUnit 3, there is not a choice like that. The two choices are to stop cooperatively (test and teardown complete) or forcibly (whatever is currently running is cancelled). I'm not sure, but I think we would need to allow more options in the engine in order to emulate exactly how V2 works. @ChrisMaddock does that seem right to you?

**Comment by CharliePoole** [comment]
I should have said that V2 worked as described __in some situations__. As with NUnit 3, it would depend on what was running when you clicked the button. If your tests are (as usual) a lot longer-running than your teardown, then the odds are that a test is running. Also, tests are more likely to be stuck in a loop.

We should discuss this to see if any improvements should be made before release or if we should try to design something to be implemented with 1.1.

**Comment by MartinZec** [comment]
Are there any improvements planned in 1.1?

**Comment by CharliePoole** [comment]
Hmm... This issue has not been popping up in my planning because I never assigned any labels to it, which I just did. As indicated, it's dependent on changes to the engine, at least for a complete fix. There may be a workaround we can implement in the interim.

The 1.1 release was about to go out a few days ago when we found some problems. Once they are fixed I'd like to release rapidly. I'll take a look at this to see if there is some sort of quick and dirty workaround I can get in.

Note that it's simply __not possible__ to do what NUnit V2 was able to do: kill the test but still run the teardown. Only the framework itself is close enough to the action to do that and it's not how the NUnit 3 framework works. Consequently, we have to either wait for the test to cancel cooperatively or kill it pretty much blindly, without knowing whether we are stopping a hung test or a hung teardown. This will leave some corner cases (small, I hope) that can't be dealt with completely.

If I can't get anything into 1.1, there will be some point releases before 1.2 and we may be able to get a fix in there. In the short term, after 1.1, I want to work on automating the release process so that I can get them out more frequently! (Looking for people to help with all of this, BTW)

**Comment by MartinZec** [comment]
Waiting for the test to cancel cooperatively does not work in my case since there is an infinite loop.
The only way would be killing it blind.
In my point of view after a test is done/finished/failed/killed, the teardown should clean up, no matter what the result was. So also in the case a test was killed.

Will the 1.1.1 release address this issue?

**Comment by CharliePoole** [comment]
Now that we are no longer dependent on the NUnit Engine, there may be some options - I'll take a look and get back to you...

**Comment by MartinZec** [comment]
Hello,

Are there any other options to trigger a teardown and cleanup when stop button is pressed?

Best regards,
Martin

**Comment by CharliePoole** [comment]
@MartinZec At the moment, the engine doesn't allow you to have both a guaranteed shutdown and and cleanup. The GUI could give you a choice between the two, but it currently doesn't. Either engine or gui could (but do not) implement an escalation strategy, where timeouts were allowed to run for a while but then a hard cancel would be issued if the run did not terminate within a reasonable period.

This issue has been languishing for a while while I was working on other projects. I'm increasing the priority and I'll try to come up with and implement some strategy. I'd like to hear any suggestions you or others have for what's the best approach.

Questions to be resolved:

* Give the user a choice or implement an automatic escalation strategy?

* If we give the user a choice, how should the GUI be changed so they can express it.

* If we have an escalation strategy,
  - how should it work?
  - should it be in the engine or the GUI?

* In either case, are there settings we want to allow the user to change, such as timing?

**Comment by MartinZec** [comment]
In my point of view there should be an automatic escalation strategy stopping test execution, calling TearDown and OneTimeTearDown after a configurable timeout.

**Comment by CharliePoole** [comment]
@MartinZec Let's spell that out a bit more, as it seems to involve some complications.

I believe you are suggesting the following...

1. User presses cancel button.

2. The GUI uses the engine's cooperative (non-forced) StopRun to cancel the run.

3. The GUI waits for the test run to complete, using a timeout.

4. If the timeout expires, we run all the teardowns and onetime teardowns.

Is that what you meant?

The difficulty with  4 is that we (either in the engine or the GUI) know which tests have not completed but we don't know where each of them is in their execution. They may be running setups, the test code itself or teardowns. It seems as if we would want to do different things in each case. That info is inside the framework - whatever framework we are running.

If item 4 were replaced with doing a forced StopRun and killing whatever process or thread is running tests then we could guarantee termination but I'm not sure it would resolve your use case.

**Comment by CharliePoole** [comment]
Writing the last comment, it occurred to me that there are two distinct use cases where the user presses the cancel button...

1. They started a run and realized that something is wrong... e.g. another file needs to be saved.

2. The test run is hung.

Maybe we need to think of having the equivalent of two buttons.

**Comment by MartinZec** [comment]
Yes, that is what I meant.

For the difficulties in 4:
First of all: I think all not-executed and all finished tests should be ignored by the stop procedure.

For me the teardown should cleanup and freeing allocated resources in a safe way (it should not hang, errors should be caught). While the test itself has the possibility to hang/loop indefinitely.

A behavior which would make sense to me:
In case the test is in setup phase, it should be stopped, the test itself should not be called and instead the teardown.
In case of a test code is running, the test should be stopped and afterwards the teardown.
In case the teardown is running the it should be stopped. There is also the other possibility to call teardown again. This might not resolve the issue of a hang and might still hang since it was hanging before, but the handling would be the same for all three cases: stop whatever is executed and call teardown to clean-up.

The two different use cases might be resolved by 
"Cancel" -> nothing is stuck, test execution is finishing the last test and stops afterwards
"Stop" -> something is hung, it need to be stopped and clean up afterwards with teardown

**Comment by CharliePoole** [comment]
There's stuff the GUI or Engine can do and other stuff that's in the domain of the framework, which is abstracted by the driver.

"I think all not-executed and all finished tests should be ignored by the stop procedure."

The NUnit Framework does this and I think we can safely require all frameworks to do it.

"In case the test is in setup phase, it should be stopped, the test itself should not be called and instead the teardown.
In case of a test code is running, the test should be stopped and afterwards the teardown.
In case the teardown is running it should be stopped."

This happens in the framework, so it could depend on what framework is running tests. That said, this is exactly how the NUnit frameworks operate now. There's one subtle difference. NUnit V2 and early versions of NUnit 3 run all teardowns. I believe later versions only run teardowns where the corresponding (same level and class) setup was initiated. At least I speced it that way before I left the project.

"There is also the other possibility to call teardown again. This might not resolve the issue of a hang and might still hang since it was hanging before, but the handling would be the same for all three cases: stop whatever is executed and call teardown to clean-up."

This would also happen in the framework. The difference from the preceding items is that there's no way to tell the framework to do that. That means the case where the hangup is in TearDown itself requires a modification to the framework. Up to now, we have considered that part of the burden on the user: TearDowns have to be written so it doesn't happen.

I think I'll try to find or create an issue in the framework project to link to this and let the framework guys mull it over.

**Comment by MartinZec** [comment]
I am currently using latest NUnit framework 3.12.
With this version there is no teardown called after the stop button is stopping a test which is stuck in test code phase.

According to your last comment, this could be resolved by using an older/earlier framework version?
I tried to roll back to older NUnit framework version, only after 3.4.0 my project compiles. For version 3.4.0 I see the same behavior that stopping a test case where a setup was done is not calling teardown.

**Comment by CharliePoole** [comment]
@MartinZec 

No version of NUnit 3 has ever used cooperative shutdown (the form of shutdown in which teardowns are called. NUnit V2 used that sort of shutdown but gave you no way to force shutdown if the test run continued to hang. Obviously, you need both forms of shutdown - at least it seems obvious today. :smile: 

I guess my last comment was worded ambiguously. I edited it to make it clear that what you are asking - to run teardowns for everything that has not terminated correctly, even those tests whose teardowns previously hung - is not currently possible in the framework. Personally, I'm not sure that's a good idea but it's really up to the framework team to review it and decide.

Haven't had time to look for a related issue in the framework but I'll do that today.

**Comment by CharliePoole** [comment]
@MartinZec 
I've researched issues in the framework project. There are a number of them related to how teardown works, particularly as related to user cancellation and timeout. No existing runner even tries to exercise the capabilities that already exist, including the TestCentric GUI. At this point, I'm thinking we should try to use what is already there. Then, if that shows a deficit in comparison to what is is needed, we can file an issue to cover it.

So, what's already in the framework is the two kinds of cancellation, forced and "cooperative" or non-forced. Summing up previous discussion...

* The engine (both the nunit engine and testcentric's modified engine) can use either one depending on how it's called.
* The console runner doesn't exercise either approach although it kills any outstanding processes before it shuts down.
* The GUI runner only uses the forced cancellation, which is why your teardowns don't run.
* If the GUI switched to non-forced cancellation, it would continue to wait for any hung test to shut down.
* We could implement a hybrid approach: try cooperative shutdown, wait for a timeout, then do forced shutdown.

That hybrid approach is what I propose doing. We can obviously predict that it won't satisfy 100% of your needs, but it should be a significant improvement. At least initially, I'd do this in the GUI itself where it's easy to experiment and only later consider pushing it into the engine. This should at least trim several problems down to a single remaining problem - that of re-running teardowns - and we could then decide whether to push it further.

Does this make sense to you?

**Comment by MartinZec** [comment]
For me the console runner approach, to kill all outstanding processes, sounds good, as well as the hybrid approach you proposed.

**Comment by CharliePoole** [comment]
@MartinZec 
cc: @ChrisMaddock since this may eventually impact the NUnit engine.

Here's a general plan and a first cut build you can try.

#### The Overall Plan

1. Start by doing the best job possible changing __only__ the GUI itself.
2. Consider potential changes in the engine (*)
    2.1 Moving some changes from item 1 to the engine
    2.2 Further functional improvements which might only be possible in the engine. (We'll see)
3. Propose changes to the framework to the NUnit team if we think they are needed. Once made, iterate from 1 or 2.

(*) Important note: The GUI has its own engine with the same API as NUnit's. We cannot change the API. Further, we are trying to converge the two engines, so even non-API changes need discussion within the NUnit Project engine team - of which I'm a member. What this means practically is that only minor engine chagnes can be made __quickly__.

#### First Steps

I have created a GUI build that uses the non-forced `StopRun` call initially, then escalates. It seems to work fairly well.

* I removed the confirmation dialog immediately after you click stop. After all, you __did__ click it!
* A progress message is displayed while waiting for the tests to end.
* If the tests don't terminate in 15 seconds, a dialog appears offering the choice of a forced cancellation or continuing to wait. You can decide which to choose by examining which tests have terminated and which have not. All tests terminating up to this point will have their teardowns executed.
* If you continue to wait, the dialog reappears every 30 seconds.
* If you select forced cancellation, the framework is told to kill the thread. No teardowns - or no further teardowns - will be  run.

You can get this build at https://ci.appveyor.com/project/CharliePoole/testcentric-gui/builds/36578096/artifacts The zip package is probably the easiest to install.

**Comment by ChrisMaddock** [comment]
I like the GUI implementation. We've talked about a similar issue before with the adapter - I'm not sure I'm too clear on how the implement something like this in the engine which will work for different types of runners, given the different interfaces available. For example - what you've done here is great for the GUI, but wouldn't be viable in the VS adapter or console runner. 

Purely for interest sake: this is what ReSharper's GUI does, iirc. 
1. Clicking stop button changes the button icon to a "force stop" icon.
2. If the tests terminate naturally, stop icon resets to normal
3. If not, the second click of the stop button implements the force stop.

I do find that one quite intuitive. I think ReSharper isn't very dialog-oriented in general - so different design ideas, really. 🙂

**Comment by CharliePoole** [comment]
@ChrisMaddock I like the button change idea. I'll experiment with that as an alternative option. Of course, the underlying policy is the same in either case: start with cooperative stop and only force it when the user says to do so in some way or another.

If we wanted to put it in the engine, it would probably require redesign to be event-driven.

**Comment by MartinZec** [comment]
I tested version https://ci.appveyor.com/project/CharliePoole/testcentric-gui/builds/36578096/artifacts:
It behaves as you described: "No teardowns - or no further teardowns - will be run."

That is exactly the issue I am facing: no teardown is called to clean up after cancelling the test and there are still processes of the previous test running, as explained below.

In my point of view the teardown of the cancelled/aborted test should be called since setup was also called to clean up allocated resources.
In case no teardown is called, all opened/outstanding processes should be killed.

Maybe I should be more precise regarding the issue I am facing:
I mentioned also in #139, I open a COM port, do a test which does not terminate (while(1)) and the COM port is not released after killing the test. In case I want to execute the test again, the COM port is blocked since the previous test instance has it still open. The COM port was opened in Setup phase and would be closed in teardown phase.
A solution would be either calling teardown or killing all opened processes.

I thought the GUI will kill any outstanding processes, does an open COM port, opened via C# in setup, not count as an outstanding process?

**Comment by CharliePoole** [comment]
@MartinZec As I tried to stress, I'm doing this incrementally.  In this iteration, I use cooperative stop first, which should allow all non-affected tests to terminate. I then use forced stop, which tells the framework to kill threads. Nothing is done on a process level. You can help me best by evaluating each iteration on it's own terms, i.e. what it is doing right and what you think needs to be added next.

Questions for you:
1. Are the non-hung tests being treated correctly from your point of view? All should terminate correctly with their teardowns called and no new tests should be initiated if you do this while there are still tests to run.
2. Is your hung test being cancelled? Is it's process terminating normally?

Earlier we had talked about killing the process but in my tests that has so far not proven necessary because my test process terminates normally. I could add code to verify that the process terminates but that would only apply to the __test process__. Is that the process you're talking about or is it some process you have created and are communicating with?

**Comment by CharliePoole** [comment]
Well here's a bummer. My test of this contains an infinite loop and I thought it was working. But by adding some output lines, I discovered that the test keeps running. Here's the test...

```c#
[Test]
public void HangingTest()
{
    TestContext.Progress.WriteLine("Test starting");
    while (true)
    {
        Sleep(3000);
        TestContext.Progress.WriteLine("Test continuing");
    }
}
```

The framework is called to cancel the test using `StopRun(true)` but the test keeps producing output!!! I filed issue /nunit/nunit#3682 about this. So long as the framework doesn't actually cancel the hung test, I can't see what the engine or runner can do. However, @MartinZec  if you have any Ideas, I'm willing to try them.

What makes this tricky is that we have never gotten any notices when tests are successfully cancelled, so I have had to simply assume that the StopRun call works. Ideally we need both things: the call should cancel the test and we should get a notice.

**Comment by CharliePoole** [comment]
Until this is fixed in the framework, there's nothing left to do but kill the process. That means it will only work if we are running out of process. However, I don't think that's a bit drawback. That will be my next iteration.

**Comment by MartinZec** [comment]
Maybe it helps to let you know my current 3 possible workarounds:
- close testcnetric GUI and re-open 
- call the test with nunit3-console, stop it with ctrl-c
- enable "Reload before every testun" in testcentric GUI

**Comment by CharliePoole** [comment]
Of course, all of those actions kill the process. Reload in the GUI is currently implemented as save some info, unload, load, reapply saved info. 

Since the GUI is aimed at working with multiple test frameworks, I'm now taking the viewpoint that the framework may fail to implement cancel. In fact, the current framework bug means that NUnit is in that category for the moment. So the GUI should protect itself and do more to cancel if necessary. 

There are two main problems here: 
1. How to determine if doing more is necessary.
2. How to leave things in a way that you can press run again.

I should have something you can try again in a day or two.

**Comment by CharliePoole** [comment]
@MartinZec 

You can get the second iteration toward a solution here: https://ci.appveyor.com/project/CharliePoole/testcentric-gui/builds/36626956/artifacts

This works just like the first iteration, except...

1. The dialog box stops appearing once you select forced cancellation - it makes no sense after that, since there is nothing else left to do.

2. When doing a forced cancellation, all runners are now unloaded. This implies terminating any test processes and unloading any test AppDomains. Note that the termination is still cooperative: we __ask__ the process to terminate. We may need to actually kill processes eventually, but it's better if they terminate themselves so they can clean up.

In testing this, you may want to make a copy of my "hung-tests" project from the issue-374 branch. It assists in testing out the basics, which have to work before we can deal with anything more complex. I'm open to suggestions as to what should be added to the test assembly.

**Comment by MartinZec** [comment]
@CharliePoole
I tried your latest version, it worked for me and solves the reported issue.

After the run was "forcibly canceled" all processes (also the COM port in my example) were terminated and I could run the test again after forcing it to cancel. So the test was canceled in a clean way.

In my point of view it would be convenient to cancel the timeout in case the Stop button is pressed a second time, similar as suggested by @ChrisMaddock.

Thank you for you effort. I am looking forward to release 1.5 :)

**Comment by CharliePoole** [comment]
@MartinZec 
Good to hear! Thanks for your help with this.

I still have some cleanup to do and then I'll merge it in. It has been a while since I released, so once this is merged I'll do 1.5.

**Comment by CharliePoole** [comment]
@MartinZec @ChrisMaddock @mikkelbu 

At the suggestion of @ChrisMaddock I implemented an alternative user interface for test cancellation. I have a Stop button and a Force Stop button, both located at the same position in the GUI main window. I alternatively display and hide each button. When you click Stop, the Force Stop button appears. There's no time delay waiting for a Dialog to pop up. I ended up incorporating a brief (400ms) delay using Sleep because Force Stop was being triggered accidentally when it appeared too fast. The delay is short enough that there's no noticeable effect and the GUI continues to update. Sleep(1) might have worked as well.

I'd like to get some feedback on the value of the two user interfaces. The new one is found at https://ci.appveyor.com/project/CharliePoole/testcentric-gui/builds/36653249/artifacts

**Comment by MartinZec** [comment]
@CharliePoole
Also this version works fine for me. I like/prefer the Force Stop option without any dialog.

For the cooperative termination, is it possible to call the teardown after cancelling the test with Force Stop? This would be the cleanest way in my point of view.

Is there a delay for forced stop? It takes for my test about 5 seconds until the test finally is terminated and I can start it again.

**Comment by CharliePoole** [comment]
@MartinZec
I like the version without the dialog better as well. When I run all the tests in my hung-tests assembly together (fixtures in parallel) I can watch the status of the normal completion and then click Force Stop when I like.

I'm not sure what you mean by "cooperative termination" - that is, it seems to mean something different than the way I have been using the term. Currently, the initial Stop is "cooperative" in the sense that the engine waits for the framework, which cooperates by shutting down. Forced termination is used if the framework can't or won't cooperate. However, there is no "cooperation" with your individual tests, if that's you intention.

There is no way to tell the framework to run teardowns. That would require a change to the framework. Also, it would have to be told to do it at the time the cancel call were made because it would otherwise be too late to do anything else. This could be  done in several ways:

 * Having a third type of stop call rather than just forced and non-forced,
 * Passing a setting when the run starts
 * Having a new attribute annotation on the test in question

We would have to ask the framework team about that. Realistically, I have to point out that the existing forced shutdown is not even working correctly at this time and I imagine the framework team will want to fix that before implementing any enhancements. In any case, it's even possible that your problem with OneTimeTearDown will disappear if the current bug is fixed.

The forced stop doesn't include any intentional delay but it's doing a lot of work to simulate correct functioning of the forced stop in the framework. It generates notices to the GUI for each test, fixture, suite or namespace, which is still running. These must be invoked on the UI thread, of course. It usually takes a few seconds even in my simple test assembly and the time will necessarily increase as more tests are being cancelled. Much shorter than an infinite loop, however. :smiley:

**Comment by CharliePoole** [comment]
@MartinZec 
My mistake! Of course there's a delay. We have to wait some amount of time to see whether the forced stop in the framework terminates the test run. So there has to be some delay and as it happens it's 5 seconds.

**Comment by CharliePoole** [comment]
I'm going with the "Two-button" approach to trigger the stoprun. rather than the dialog-based one. It's much easier to use.

Note that this is only implemented for the Standard GUI right now. The Experimental GUI continues to use forced cancel exclusively. It will require some UI work to add controls for selecting alternative approaches. I'll create an issue for that.

**Comment by CharliePoole** [comment]
This change has now been merged into master, which means it's available from our development feed on MyGet: https://www.myget.org/feed/testcentric/package/nuget/testcentric-gui/1.4.2-dev00023

It will be released in a week or so as part of the version 1.5 release.
